src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeMemory.Unix.cs

```diff
@@ -0,0 +1,151 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>This class contains methods that are mainly used to manage native memory.</summary>
+    public static unsafe class NativeMemory
+    {
+        /// <summary>Allocates an aligned block of memory of the specified size and alignment, in bytes.</summary>
+        /// <param name="byteCount">The size, in bytes, of the block to allocate.</param>
+        /// <param name="alignment">The alignment, in bytes, of the block to allocate. This must be a power of <c>2</c> greater than or equal to <c>sizeof(void*)</c> for portability.</param>
+        /// <returns>A pointer to the allocated aligned block of memory.</returns>
+        /// <exception cref="ArgumentException"><paramref name="alignment" /> is not a power of two or is not a multiple of <c>void*</c>.</exception>
+        /// <exception cref="OutOfMemoryException">Allocating <paramref name="byteCount" /> of memory with <paramref name="alignment" /> failed.</exception>
+        /// <remarks>
+        ///     <para>This method allows <paramref name="byteCount" /> to be <c>0</c> and will return a valid pointer that should not be dereferenced and that should be passed to free to avoid memory leaks.</para>
+        ///     <para>This method is a thin wrapper over the C <c>aligned_alloc</c> API or a platform dependent aligned allocation API such as <c>_aligned_malloc</c> on Win32.</para>
+        ///     <para>This method is not compatible with <see cref="Free" /> or <see cref="Realloc" />, instead <see cref="AlignedFree" /> should be called.</para>
+        /// </remarks>
+        [CLSCompliant(false)]
+        public static void* AlignedAlloc(nuint byteCount, nuint alignment)
+        {
+            if (!BitOperations.IsPow2(alignment) || ((alignment % (uint)sizeof(void*)) != 0))
+            {
+                // The C standard doesn't define what a valid alignment is, however POSIX requires a power of two that is a multiple of void*
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_AlignmentOrByteCount);
+            }
+
+            // The C standard and POSIX requires size to be a multiple of alignment and we want an "empty" allocation for zero
+            void* result = Interop.Sys.AlignedAlloc(alignment, (byteCount != 0) ? (byteCount + (alignment - 1)) & ~(alignment - 1) : alignment);
```

Comment:
The aligning up of byteCount has integer overflow here. I think it should be done with proper check against overflow only when it would make POSIX unhappy.

Also, I would align up the alignment as well to make the POSIX happy (that one is not necessary on Windows).

Reply:
> Also, I would align up the alignment as well to make the POSIX happy (that one is not necessary on Windows).

You're suggesting we effectively do just `IsPow2` for validation and then do `alignment = Max(alignment, sizeof(void*))`? Would this also apply if `alignment == 0`?

> I think it should be done with proper check against overflow only when it would make POSIX unhappy.

POSIX (and the C standard) requires `size` to be a multiple of `alignment` and so it will always requires a check.

Overflow here can only happen for `byteCount > (nuint.MaxValue & ~(alignment - 1))` and will overflow to a non-power of two that is less than `alignment` and so will cause `null` to be returned as per the POSIX/C standard requirements. This would be consistent with any conceivable implementation since allocation `nuint.MaxValue & ~(alignment - 1)` should be impossible.

Reply:
> You're suggesting we effectively do just IsPow2 for validation

Yep.

> Would this also apply if alignment == 0?

IsPow2 validation should fail for `alignment == 0`.

> Overflow here can only happen for byteCount > (nuint.MaxValue & ~(alignment - 1)) and will overflow to a value that is less than alignment and so will cause null to be returned as per the POSIX/C standard requirements.

We should have it covered by tests. (A comment about why this overflow is harmless may be useful too.)

Reply:
If we are willing to trust that posix alligned alloc does the argument validation and returns null as speced, you can do the managed check after the fact - only once the posix alligned alloc returns null.

Reply:
> you can do the managed check after the fact

Windows requires the `IsPow2` check first, since it calls the hookable invalid argument handler and we don't want that.
We could conceivable move the check after for Unix, but that changes the assumptions we can make about the upsizing checks since the the code doing that assumes/requires power of two for `alignment` and so I think its better to do it first.

Reply:
Sounds good.

Do you still plan to do the fixup of the alignment to make Posix happy? And remove `sizeof(void*)` check for Windows since the Windows API should work fine for any alignment.

Reply:
Yes, I actually had done that already and simply didn't push it.